policy: Add Clang thread safety annotations for variables guarded by cs_feeEstimator

practicalswift opened the pull request:
* Add Clang thread safety annotations for variables guarded by `cs_feeEstimator`
* ~~Add missing `cs_feeEstimator` locks~~

MarcoFalke commented on src/policy/fees.h:
```diff
@@ -230,10 +230,12 @@ class CBlockPolicyEstimator
-    unsigned int firstRecordedHeight;
-    unsigned int historicalFirst;
-    unsigned int historicalBest;
+    mutable CCriticalSection cs_feeEstimator;
```
Might as well rename to `m_cs_fee_estimator` according to developer notes.

practicalswift commented:
@MarcoFalke Good idea. Now rename from `cs_feeEstimator` to `m_cs_fee_estimator`. Please re-review :-)

Empact commented on src/policy/fees.cpp:
```diff
@@ -710,6 +711,7 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
 
 unsigned int CBlockPolicyEstimator::HighestTargetTracked(FeeEstimateHorizon horizon) const
 {
+    LOCK(cs_feeEstimator);
```
I don't think we need exclusive locking for these operations, as `GetMaxConfirms` so concurrent access wouldn't result in incoherent data.

Empact commented on src/policy/fees.cpp:
```diff
@@ -672,6 +672,8 @@ CFeeRate CBlockPolicyEstimator::estimateFee(int confTarget) const
 
 CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThreshold, FeeEstimateHorizon horizon, EstimationResult* result) const
 {
+    LOCK(cs_feeEstimator);
```
`get` is a read of the ptr value, which is only updated in the constructor, so this lock is unnecessary AFAICT. `PT_GUARDED_BY` could be more appropriate.

Empact commented on src/policy/fees.cpp:
```diff
@@ -710,6 +711,7 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
 
 unsigned int CBlockPolicyEstimator::HighestTargetTracked(FeeEstimateHorizon horizon) const
 {
+    LOCK(m_cs_fee_estimator);
```
I don't think we need exclusive locking here, as `GetMaxConfirms` is a read so concurrent access wouldn't result in incoherent data.

practicalswift replied:
@Empact Do you mean that the following annotations are incorrect and should be change or removed?

```
    std::unique_ptr<TxConfirmStats> feeStats GUARDED_BY(m_cs_fee_estimator);
    std::unique_ptr<TxConfirmStats> shortStats GUARDED_BY(m_cs_fee_estimator);
    std::unique_ptr<TxConfirmStats> longStats GUARDED_BY(m_cs_fee_estimator);
```

Empact commented on src/policy/fees.cpp:
```diff
@@ -672,6 +672,8 @@ CFeeRate CBlockPolicyEstimator::estimateFee(int confTarget) const
 
 CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThreshold, FeeEstimateHorizon horizon, EstimationResult* result) const
 {
+    LOCK(m_cs_fee_estimator);
```
`get` is a read of the ptr value, which is only updated in the constructor, so this lock is unnecessary AFAICT. `PT_GUARDED_BY` could be more appropriate.

practicalswift committed:
policy: Add Clang thread safety annotations for variables guarded by cs_feeEstimator

practicalswift committed:
scripted-diff: Rename from cs_feeEstimator to m_cs_fee_estimator

-BEGIN VERIFY SCRIPT-
sed -i 's/cs_feeEstimator/m_cs_fee_estimator/' src/policy/fees.cpp src/policy/fees.h
-END VERIFY SCRIPT-

practicalswift commented:
@Empact @MarcoFalke Updated. Would you mind re-reviewing? The current version is annotations only, so no change in locking. In other words: this PR changes only compile-time checking and no runtime behaviour. Should hence hopefully be trivial to review :-)

ken2812221 commented:
utACK 1a01e1c120809b19ddc9c9c6ee5d6686a90f55b3

Empact commented on src/policy/fees.h:
```diff
@@ -241,34 +243,32 @@ class CBlockPolicyEstimator
     /** Classes to track historical data on transaction confirmations */
     std::unique_ptr<TxConfirmStats> feeStats;
     std::unique_ptr<TxConfirmStats> shortStats;
     std::unique_ptr<TxConfirmStats> longStats;
```
I'm curious - did you try marking these `PT_GUARDED_BY`? Seems it would be appropriate, based on the lock in `estimateRawFee`.

practicalswift committed:
Add locking annotations to feeStats, shortStats and longStats
